packages/contracts/contracts/deploy/PayableDAppFactory.sol

```diff
@@ -0,0 +1,144 @@
+    /// @param _consensus The consensus contract to use
+    /// @param _price The price per second of execution of the dapp node
+    constructor(
+        address _owner,
+        ICartesiDAppFactory _factory,
+        IERC20 _token,
+        IConsensus _consensus,
+        uint256 _price
+    ) {
+        factory = _factory;
+        token = _token;
+        consensus = _consensus;
+        price = _price;
+
+        // transfer ownership to the give owner
+        transferOwnership(_owner);
+    }
+
+    /// @dev This function deploys a new application, and registers it with the payment system
+    function newApplication(
+        address _dappOwner,
+        bytes32 _templateHash,
+        string memory _cid,
+        uint256 _runway
+    ) external override returns (CartesiDApp) {
+        // delegate call to factory
+        CartesiDApp dapp = factory.newApplication(
+            consensus,
+            _dappOwner,
+            _templateHash
+        );
+
+        // emit event to bind CID to application
+        emit MachineLocation(address(dapp), _cid);
+
+        // set initial runway
+        _extendRunway(msg.sender, dapp, _runway);
+
+        // return application
+        return dapp;
+    }
+
+    function importApplication(
+        CartesiDApp _dapp,
+        string memory _cid,
+        uint256 _runway
+    ) external {
+        require(
+            _dapp.getConsensus() == this.consensus(),
+            "PaymentDAppFactory: cannot import dapp with a different consensus"
+        );
+
+        require(
+            _dapp.owner() == msg.sender,
+            "PaymentDAppFactory: only owner can import application"
+        );
+
+        // emit location event, as DApp was not intantiated by this factory
+        emit MachineLocation(address(_dapp), _cid);
+
+        // extend runway
+        _extendRunway(msg.sender, _dapp, _runway);
+    }
+
+    function cost(uint256 _time) external view returns (uint256) {
+        return _time * price;
+    }
+
+    function extendRunway(
+        ICartesiDApp _dapp,
+        uint256 _time
+    ) external returns (uint256) {
+        return _extendRunway(msg.sender, _dapp, _time);
+    }
+
+    function _extendRunway(
+        address _msgSender,
+        ICartesiDApp _dapp,
+        uint256 _time
+    ) private returns (uint256) {
+        // calculate the token cost of the runway time
+        uint256 _cost = this.cost(_time);
+
+        // transfer tokens from sender to this contract
+        require(
+            token.transferFrom(_msgSender, address(this), _cost),
+            "PaymentDAppFactory: failed to transfer tokens"
+        );
+
+        uint256 currentRunway = runway[_dapp];
+
+        // if runway was in the past, update to now
+        currentRunway = currentRunway < block.timestamp
+            ? block.timestamp
+            : currentRunway;
+
+        // update dapp runway
+        uint256 newRunway = currentRunway + _time;
+        runway[_dapp] = newRunway;
+
```

Comment:
Here would be a good place to have the maximum runway check.
Something like:  `require( newRunway <= block.timestamp + 1 month, "can't extend until more than one month in the future")`.

I don't know what the conclusion was, but I think in our conversation we were thinking that it made sense to limit the max runway

Reply:
We talked about that because the customer can't withdraw.
The validator can withdraw and refund the customer though, in a friendly relationship.

Reply:
True.  However, I think it might make accounting harder offchain. You never know whats the amount of funds that you might have to give back. I also think it makes the price raising thing a bit more complicated, even if you price raise by migrating.

But I see your point!